### ISSUE TITLE
Another way to train G and D network

### ISSUE THREAD

**Issue by John1231983** [hfslyc/AdvSemiSeg#15]
Hello, in your implementation, you first train G network, then train on D network. The code can be summarized as
```
#=========Train on G=========
 for param in model_D.parameters():
                param.requires_grad = False
loss_seg = loss_calc(pred, labels, args.gpu)
loss_adv_pred = bce_loss(D_out, make_D_label(gt_label, ignore_mask))
loss = loss_seg + args.lambda_adv_pred * loss_adv_pred
loss.backward()
#=======Train on D===========
for param in model_D.parameters():
                param.requires_grad = True
pred = pred.detach()
#-----Fake-------
loss_D = bce_loss(D_out, make_D_label(pred_label, ignore_mask))
loss_D.backward()
#-----Real--------
loss_D = bce_loss(D_out, make_D_label(gt_label, ignore_mask_gt))
loss_D.backward()
optimizer.step()
optimizer_D.step()
```
However, follow the DCGAN example, I make more simple way that does not requires setting `param.requires_grad = False/True`. This is my modified

```
#=======Train on D===========
#-----Real--------
loss_D = bce_loss(D_out, make_D_label(gt_label, ignore_mask_gt))
loss_D.backward()
#-----Fake-------
pred= model(image)#add more here
pred = pred.detach()
loss_D = bce_loss(D_out, make_D_label(pred_label, ignore_mask))
loss_D.backward()
optimizer_D.step()
#=========Train on G=========
loss_seg = loss_calc(pred, labels, args.gpu)
loss_adv_pred = bce_loss(D_out, make_D_label(gt_label, ignore_mask))
loss = loss_seg + args.lambda_adv_pred * loss_adv_pred
loss.backward()
optimizer.step()
```

In this modification, I trained the D network first with real label and `detach` for the fake label. We did not use  `param.requires_grad = False/True` because `detach` did it. The code is enabled to run. However, I did not guarantee it equivalent with your purpose or not "During the training G network, we fix the D network`. Could you give me some comment?  

Reference :https://discuss.pytorch.org/t/when-training-gan-why-do-we-not-need-to-zero-grad-discriminator/20439
>In the update step of the discriminator (line 208), the generator does not get the data, so the backward step does not calculate any gradients for it.
In line 217 the input to the discriminator is detached as you already observed. Thus the backward call of errD_fake also does not calculate the gradients for the generator.
Before updating the generator (line 225 etc.) the gradients are zeroed, so it looks alright

**Comment by John1231983** [comment]
@hfslyc: please look at the problem

**Comment by hfslyc** [comment]
Sorry I was busying on something and forget to reply.

Yes, I think your code looks good. However, since we are using the same image pairs for training D and G (in DCGAN they use two data loaders), there is no pred to train the D first. If you want to reuse the pred from the previous iteration or forward a new batch, the code would just become messier or less efficient in my opinion.

In our environment, disabling gradient does not take much time. So I would prefer to let the code stay at what it is now. Also, disabling them explicitly makes the code more self-explanatory.

**Comment by John1231983** [comment]
Really. I checked in dcgan of pytorch example. They used one dataloader for both G and D. While you used two dataloaders. Something typo in your comment?

**Comment by hfslyc** [comment]
I was referring to the original dcgan torch implementation. yeah I think you are correct. But I would prefer not to update the code for better readability.
